apps/lib/chain_pusher/solana.go

```diff
@@ -0,0 +1,411 @@
+				}
+
+				data := msg.Value.Data.GetBinary()
+
+				decoder := bin.NewBorshDecoder(data)
+
+				account := &contract.TemporalNumericValueFeedAccount{}
+
+				err = account.UnmarshalWithDecoder(decoder)
+
+				if err != nil {
+					sci.logger.Error().Err(err).Msg("Error getting account from message")
+					continue
+				}
+
+				latestValue := account.LatestValue
+				tv := InternalStorkStructsTemporalNumericValue{
+					QuantizedValue: latestValue.QuantizedValue.BigInt(),
+					TimestampNs:    latestValue.TimestampNs,
+				}
+
+				ch <- map[InternalEncodedAssetId]InternalStorkStructsTemporalNumericValue{account.Id: tv}
+			}
+		}(sub, feedAccount)
+	}
+
+	// Wait indefinitely
+	select {}
+}
+
+func (sci *SolanaContractInteracter) PullValues(encodedAssetIds []InternalEncodedAssetId) (map[InternalEncodedAssetId]InternalStorkStructsTemporalNumericValue, error) {
+	polledVals := make(map[InternalEncodedAssetId]InternalStorkStructsTemporalNumericValue)
+
+	for _, encodedAssetId := range encodedAssetIds {
+		// Derive the PDA for this asset
+		feedAccount, _, err := solana.FindProgramAddress(
+			[][]byte{
+				[]byte("stork_feed"),
+				encodedAssetId[:],
+			},
+			sci.contractAddr,
+		)
+		if err != nil {
+			sci.logger.Error().Err(err).Str("assetId", hex.EncodeToString(encodedAssetId[:])).Msg("Failed to derive PDA for feed account in PullValues")
+			continue
+		}
+
+		// Fetch the account data
+		accountInfo, err := sci.client.GetAccountInfo(context.Background(), feedAccount)
+		if err != nil {
+			sci.logger.Error().Err(err).Str("account", feedAccount.String()).Msg("Failed to get account info")
+			continue
+		}
+
+		if accountInfo == nil || len(accountInfo.Value.Data.GetBinary()) == 0 {
+			sci.logger.Debug().Str("assetId", hex.EncodeToString(encodedAssetId[:])).Msg("No value found")
+			continue
+		}
+
+		// Decode the account data
+		decoder := bin.NewBorshDecoder(accountInfo.Value.Data.GetBinary())
+		account := &contract.TemporalNumericValueFeedAccount{}
+		err = account.UnmarshalWithDecoder(decoder)
+		if err != nil {
+			sci.logger.Error().Err(err).Str("account", feedAccount.String()).Msg("Failed to decode account data")
+			continue
+		}
+
+		// Convert to internal format
+		polledVals[encodedAssetId] = InternalStorkStructsTemporalNumericValue{
+			QuantizedValue: account.LatestValue.QuantizedValue.BigInt(),
+			TimestampNs:    account.LatestValue.TimestampNs,
+		}
+	}
+
+	return polledVals, nil
+}
+
+func (sci *SolanaContractInteracter) BatchPushToContract(priceUpdates map[InternalEncodedAssetId]AggregatedSignedPrice) error {
+	// Create a rate limiter: 40 requests per 10 seconds (4 per second)
+	// with a burst of 10 to stay well within the limits
+	limiter := rate.NewLimiter(rate.Every(250*time.Millisecond), 10)
+
+	var wg sync.WaitGroup
+	errChan := make(chan error, len(priceUpdates))
+
+	priceUpdatesBatch := make(map[InternalEncodedAssetId]AggregatedSignedPrice)
+	i := 0
+	for encodedAssetId, priceUpdate := range priceUpdates {
```

Comment:
Would also make this piece of it testable which would be nice

Reply:
I actually initially wrote batching and pushing as two separate loops, and then went and did it this way so that it'd be faster. I can break out batching into a separate function, I agree readability and test ability would improve, and I'd imagine that the performance hit would be negligible.